README.md
# Bull Prom
[![npm version](https://badge.fury.io/js/bull-prom.svg?style=flat)](http://badge.fury.io/js/bull-prom)

Provides [Prometheus](https://prometheus.io/) metrics for [Bull](https://github.com/OptimalBits/bull)

## Metrics

| Metric                              | type    | description                                             |
|-------------------------------------|---------|---------------------------------------------------------|
| jobs_completed_total                | counter | Total number of completed jobs                          |
| jobs_active_total                   | counter | Total number of active jobs (currently being processed) |
| jobs_delayed_total                  | counter | Total number of jobs that will run in the future        |
| jobs_failed_total                   | counter | Total number of failed jobs                             |
| jobs_waiting_total                  | counter | Total number of jobs waiting to be processed            |
| jobs_duration_milliseconds          | summary | Processing time for completed/failed                    |
| jobs_waiting_duration_milliseconds  | summary | Waiting time for completed/failed                       |
| jobs_attempts                       | summary | Processing time for completed/failed/jobs               |

_IMPORTANT_: If you are using this library to track job duration metrics for a queue for multiple consumers you will need to listen to the `global:completed` event. Otherwise this library will not record job duration metrics. To listen to this event instead of the `completed` event you'll need to set the init paremeter `useGlobal` to true.

## Usage
```typescript
import Queue from 'bull';
import promClient from 'prom-client';
import * as bullProm from 'bull-prom';

const queue = new Queue('myQueue'...);

const bullMetric = bullProm.init({
  promClient, // optional, it will use internal prom client if it is not given
  interval: 1000, // optional, in ms, default to 60000
  useGlobal: false, // optional, default to false
});

const started = bullMetric.start(queue);

// Optional
started.stop();

// Metrics result in Prometheus
// jobs_waiting_total{queue_name="myQueue", queue_prefix="default"} 0
// jobs_active_total{queue_name="myQueue", queue_prefix="default"} 0
// jobs_complete_total{queue_name="myQueue", queue_prefix="default"} 0
// jobs_failed_total{queue_name="myQueue", queue_prefix="default"} 0
// jobs_delayed_total{queue_name="myQueue", queue_prefix="default"} 0
```

## API
### init(options)
Initialize

options:
- `promClient` (*optional*): prom client instance
- `interval` (*optional*, default `60000`): interval in ms to fetch the Bull statistic
- `useGlobal` (*optional*, default `false`)

### start(queue)
Start running and fetching the data from Bull based on interval with the given Bull queue.

Returns a queue metrics object which includes the following methods:
- `stop()`: stops monitoring the queue metrics
- `remove()`: removes metrics from prometheus


## Contributors

* **@mjgp2**
* **@robbiet480**
* **@TotallyNotElite**
* **@ejhayes**

## License
MIT © [Pawel Badenski](https://github.com/pbadenski)

This library is largely derived from [kue-prom](https://github.com/deerawan/kue-prom) (MIT © [Budi Irawan](https://github.com/deerawan))


CHANGELOG.md
<a name="3.2.1"></a>
## [3.2.1](https://github.com/pbadenski/bull-prom/compare/v3.1.1...v3.2.1) (2021-05-19)


### Features

* add configuration option to listen to global:completed event. Otherwise the local completed event will not work for completed jobs running elsewhere. ([c6a7af2](https://github.com/pbadenski/bull-prom/commit/c6a7af2693d09c9a5bdfb6910f56e61cce83ddee))



<a name="3.1.1"></a>
## [3.1.1](https://github.com/pbadenski/bull-prom/compare/v3.1.0...v3.1.1) (2021-05-11)

### Chores

* upgrade all depeendencies to latest versions

<a name="3.1.0"></a>
# [3.1.0](https://github.com/pbadenski/bull-prom/compare/v3.0.0...v3.1.0) (2021-05-11)

### Features

* Add remove() function to queue metrics to allow removing all queue specific metrics from prometheus **@ejhayes** ([854b49c](https://github.com/pbadenski/bull-prom/commit/854b49c861f5dac7471790ba4e37aa5cb6d3d641))

<a name="3.0.0"></a>
# [3.0.0](https://github.com/pbadenski/bull-prom/compare/v2.1.0...v3.0.0) (2020-06-03)

### Features

* Add new 'queue_prefix' label **@robbiet480**
* Duration summary **@mjgp2**, **@robbiet480**
* Multi-queue support **@mjgp2**

### Bug Fixes

* upgrade yarn.lock ([f04f27d](https://github.com/pbadenski/bull-prom/commit/f04f27d4132f7d1f766c5f63867a329aaf3d45b7))

### Chores

* upgrade to prom-client v12 **@robbiet480**

<a name="2.1.0"></a>
## [2.1.0](https://github.com/pbadenski/bull-prom/compare/v2.0.0...v2.1.0) (2020-05-27)

### Chores

* **dependencies**: upgrade all to latest version ([8526b1](https://github.com/pbadenski/bull-prom/commit/8526b1)), closes ([#2](https://github.com/pbadenski/bull-prom/issues/2)) **@robbiet480**


test/metric-spec.ts
import KueProm from '../src/metric';
import * as sinon from 'sinon';
import * as chai from 'chai';
import * as kue from 'kue';

const expect = chai.expect;

describe('metric', () => {
  describe('#run', () => {
    let createQueueStub;
    let inactiveStub;
    let queue;

    before(() => {
      createQueueStub = sinon.stub(kue, 'createQueue').returns(mockCreateQueue);
      queue = kue.createQueue();
      inactiveStub = sinon.stub(queue, 'inactiveCount', (name, callback) => { callback(); });
      const metric = KueProm({queue, jobName: 'lovely'});
    });
  });
});

function mockCreateQueue() {
  return {
    activeCount(name, callback) { callback(); },
    inactiveCount(name, callback) { callback(); },
    completeCount(name, callback) { callback(); },
    failedCount(name, callback) { callback(); },
    delayedCount(name, callback) { callback(); },
  };
}


test/greeter-spec.ts

import { Greeter } from "../src/metric";
import * as chai from "chai";

const expect = chai.expect;

describe("greeter", () => {
  it("should greet with message", () => {
    const greeter = new Greeter("friend");
    expect(greeter.greet()).to.equal("Bonjour, friend!");
  });
});


lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.init = exports.JobStatus = void 0;
const client = require("prom-client");
var JobStatus;
(function (JobStatus) {
    JobStatus["COMPLETED"] = "completed";
    JobStatus["FAILED"] = "failed";
})(JobStatus = exports.JobStatus || (exports.JobStatus = {}));
function init(opts) {
    const { interval = 60000, promClient = client, useGlobal = false } = opts;
    const QUEUE_NAME_LABEL = 'queue_name';
    const QUEUE_PREFIX_LABEL = 'queue_prefix';
    const STATUS_LABEL = 'status';
    const activeMetricName = 'jobs_active_total';
    const waitingMetricName = 'jobs_waiting_total';
    const completedMetricName = 'jobs_completed_total';
    const failedMetricName = 'jobs_failed_total';
    const delayedMetricName = 'jobs_delayed_total';
    const durationMetricName = 'jobs_duration_milliseconds';
    const waitingDurationMetricName = 'jobs_waiting_duration_milliseconds';
    const attemptsMadeMetricName = 'jobs_attempts';
    const completedMetric = new promClient.Gauge({
        name: completedMetricName,
        help: 'Number of completed jobs',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
    });
    const failedMetric = new promClient.Gauge({
        name: failedMetricName,
        help: 'Number of failed jobs',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
    });
    const delayedMetric = new promClient.Gauge({
        name: delayedMetricName,
        help: 'Number of delayed jobs',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
    });
    const activeMetric = new promClient.Gauge({
        name: activeMetricName,
        help: 'Number of active jobs',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
    });
    const waitingMetric = new promClient.Gauge({
        name: waitingMetricName,
        help: 'Number of waiting jobs',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
    });
    const durationMetric = new promClient.Summary({
        name: durationMetricName,
        help: 'Time to complete jobs',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL, STATUS_LABEL],
        maxAgeSeconds: 300,
        ageBuckets: 13,
    });
    const waitingDurationMetric = new promClient.Summary({
        name: waitingDurationMetricName,
        help: 'Time spent waiting for a job to run',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL, STATUS_LABEL],
        maxAgeSeconds: 300,
        ageBuckets: 13
    });
    const attemptsMadeMetric = new promClient.Summary({
        name: attemptsMadeMetricName,
        help: 'Job attempts made',
        labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL, STATUS_LABEL],
        maxAgeSeconds: 300,
        ageBuckets: 13
    });
    function recordJobMetrics(labels, status, job) {
        if (!job.finishedOn) {
            return;
        }
        const jobLabels = Object.assign({ [STATUS_LABEL]: status }, labels);
        const jobDuration = job.finishedOn - job.processedOn;
        durationMetric.observe(jobLabels, jobDuration);
        const waitingDuration = job.processedOn - job.timestamp;
        waitingDurationMetric.observe(jobLabels, waitingDuration);
        attemptsMadeMetric.observe(jobLabels, job.attemptsMade);
    }
    function start(queue) {
        const keyPrefix = queue.keyPrefix.replace(/.*\{|\}/gi, '');
        const labels = {
            [QUEUE_NAME_LABEL]: queue.name,
            [QUEUE_PREFIX_LABEL]: keyPrefix,
        };
        if (useGlobal) {
            queue.on('global:completed', (jobId) => __awaiter(this, void 0, void 0, function* () {
                const job = yield queue.getJob(jobId);
                recordJobMetrics(labels, JobStatus.COMPLETED, job);
            }));
            queue.on('global:failed', (jobId) => __awaiter(this, void 0, void 0, function* () {
                const job = yield queue.getJob(jobId);
                recordJobMetrics(labels, JobStatus.FAILED, job);
            }));
        }
        else {
            queue.on('completed', (job) => {
                recordJobMetrics(labels, JobStatus.COMPLETED, job);
            });
            queue.on('failed', (job) => {
                recordJobMetrics(labels, JobStatus.FAILED, job);
            });
        }
        const metricInterval = setInterval(() => {
            queue
                .getJobCounts()
                .then(({ completed, failed, delayed, active, waiting }) => {
                completedMetric.set(labels, (completed || 0));
                failedMetric.set(labels, (failed || 0));
                delayedMetric.set(labels, (delayed || 0));
                activeMetric.set(labels, (active || 0));
                waitingMetric.set(labels, (waiting || 0));
            });
        }, interval);
        const removeMetrics = () => {
            durationMetric.remove(labels);
            completedMetric.remove(labels);
            failedMetric.remove(labels);
            delayedMetric.remove(labels);
            activeMetric.remove(labels);
            waitingMetric.remove(labels);
            waitingDurationMetric.remove(labels);
            attemptsMadeMetric.remove(labels);
        };
        return {
            stop: () => clearInterval(metricInterval),
            remove: removeMetrics
        };
    }
    return {
        start,
    };
}
exports.init = init;


lib/metric.js
"use strict";
exports.__esModule = true;
var promClient = require("prom-client");
var KueProm = function (opts) {
    var queue = opts.queue, jobName = opts.jobName, _a = opts.interval, interval = _a === void 0 ? 60000 : _a, _b = opts.prefixMetricName, prefixMetricName = _b === void 0 ? '' : _b;
    var activeMetricName = getFullMetricName(jobName, 'active_count', prefixMetricName);
    var inactiveMetricName = getFullMetricName(jobName, 'inactive_count', prefixMetricName);
    var completeMetricName = getFullMetricName(jobName, 'complete_count', prefixMetricName);
    var failedMetricName = getFullMetricName(jobName, 'failed_count', prefixMetricName);
    var delayedMetricName = getFullMetricName(jobName, 'delayed_count', prefixMetricName);
    var activeMetric = new promClient.Gauge(activeMetricName, 'Number of active job');
    var inactiveMetric = new promClient.Gauge(inactiveMetricName, 'Number of inactive job');
    var completeMetric = new promClient.Gauge(completeMetricName, 'Number of complete job');
    var failedMetric = new promClient.Gauge(failedMetricName, 'Number of failed job');
    var delayedMetric = new promClient.Gauge(delayedMetricName, 'Number of delayed job');
    var metricInterval;
    function run() {
        metricInterval = setInterval(function () {
            queue.activeCount(jobName, function (err, total) {
                activeMetric.set(total);
            });
            queue.inactiveCount(jobName, function (err, total) {
                inactiveMetric.set(total);
            });
            queue.completeCount(jobName, function (err, total) {
                completeMetric.set(total);
            });
            queue.failedCount(jobName, function (err, total) {
                failedMetric.set(total);
            });
            queue.delayedCount(jobName, function (err, total) {
                delayedMetric.set(total);
            });
        }, interval);
    }
    function stop() {
        metricInterval.clearInterval();
    }
    return {
        run: run,
        stop: stop
    };
};
function getFullMetricName(jobName, metricName, prefixName) {
    if (prefixName === void 0) { prefixName = ''; }
    prefixName = (prefixName) ? prefixName + "_" : '';
    return "" + prefixName + jobName + "_" + metricName;
}
exports["default"] = KueProm;


src/metric.ts
import promClient = require('prom-client');
import * as kue from 'kue';

export interface Options {
  queue: kue.Queue;
  prefixMetricName: string;
  jobName: string;
}

const KueProm = (opts: Options) => {
  const {queue, jobName, prefixMetricName = ''} = opts;

  const activeMetricName = getFullMetricName(jobName, 'active_count', prefixMetricName);
  const inactiveMetricName = getFullMetricName(jobName, 'inactive_count', prefixMetricName);
  const completeMetricName = getFullMetricName(jobName, 'complete_count', prefixMetricName);
  const failedMetricName = getFullMetricName(jobName, 'failed_count', prefixMetricName);
  const delayedMetricName = getFullMetricName(jobName, 'delayed_count', prefixMetricName);

  const activeMetric = new promClient.Gauge(activeMetricName, 'Number of active job');
  const inactiveMetric = new promClient.Gauge(inactiveMetricName, 'Number of inactive job');
  const completeMetric = new promClient.Gauge(completeMetricName, 'Number of complete job');
  const failedMetric = new promClient.Gauge(failedMetricName, 'Number of failed job');
  const delayedMetric = new promClient.Gauge(delayedMetricName, 'Number of delayed job');

  function start() {
    queue.activeCount(jobName, (err: any, total: number) => {
      activeMetric.set(total);
    });

    queue.inactiveCount(jobName, (err: any, total: number) => {
      inactiveMetric.set(total);
    });

    queue.completeCount(jobName, (err: any, total: number) => {
      completeMetric.set(total);
    });

    queue.failedCount(jobName, (err: any, total: number) => {
      failedMetric.set(total);
    });

    queue.delayedCount(jobName, (err: any, total: number) => {
      delayedMetric.set(total);
    });
  }
};

function getFullMetricName(jobName: string, metricName: string, prefix: string = '') {
  const prefixName = (prefix) ? `${prefix}_` : '';

  return `${prefixName}${jobName}_${metricName}`;
}

export default KueProm;


scripts/publish.sh
#!/bin/bash

if [[ -z $(git status -uno --porcelain) ]]; then
  VERSION="$(npm view bull-prom version)";
  read -p "Enter the new version number: (currently ${VERSION}) " BUMP;
  VERSION="$(npm version $BUMP --no-git-tag-version)";
  node_modules/.bin/conventional-changelog -p angular -i CHANGELOG.md -s;
  vi CHANGELOG.md;
  git diff package.json CHANGELOG.md;
  read -p "Look good? (y/n) " CONDITION;

  if [ "$CONDITION" == "y" ]; then
    git add package.json CHANGELOG.md;
    git commit -m "chore(publish): ${VERSION}";
    git tag "${VERSION}" -m "See https://github.com/pbadenski/bull-prom/blob/master/CHANGELOG.md#${BUMP}";
    git push origin master;
    git push origin "${VERSION}";
    read -p "Which dist-tag? (latest) " DIST_TAG;
    DIST_TAG=${DIST_TAG:-latest}

    if [[ "$BUMP" =~ - ]] && [ "$DIST_TAG" == "latest" ]; then
      read -p "Using dist-tag 'latest' for a pre-release. ARE YOU SURE? (y/n) " CONDITION;

      if [ "$CONDITION" == "n" ] || [ "$CONDITION" == "" ]; then
        echo "Cancelled publish by your request!";
        exit 1;
      fi
    fi

    read -p "Enter 2FA auth token: " AUTH;
    yarn publish --tag $DIST_TAG --otp $AUTH;
  else
    git checkout -f package.json CHANGELOG.md;
    echo "Cancelled publish by your request!";
    exit 1;
  fi

else
  echo "You cannot publish with uncommited changes";
  exit 1;
fi


src/index.ts
import client = require('prom-client');
import * as bull from 'bull';

export interface Options {
  promClient?: typeof client;
  interval?: number;
  useGlobal?: boolean;
}

export enum JobStatus {
  COMPLETED = 'completed',
  FAILED = 'failed'
}

export function init(opts: Options) {
  const { interval = 60000, promClient = client, useGlobal = false } = opts;

  const QUEUE_NAME_LABEL = 'queue_name';
  const QUEUE_PREFIX_LABEL = 'queue_prefix';
  const STATUS_LABEL = 'status';

  const activeMetricName = 'jobs_active_total';
  const waitingMetricName = 'jobs_waiting_total';
  const completedMetricName = 'jobs_completed_total';
  const failedMetricName = 'jobs_failed_total';
  const delayedMetricName = 'jobs_delayed_total';
  const durationMetricName = 'jobs_duration_milliseconds';
  const waitingDurationMetricName = 'jobs_waiting_duration_milliseconds';
  const attemptsMadeMetricName = 'jobs_attempts';

  const completedMetric = new promClient.Gauge({
    name: completedMetricName,
    help: 'Number of completed jobs',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
  });

  const failedMetric = new promClient.Gauge({
    name: failedMetricName,
    help: 'Number of failed jobs',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
  });

  const delayedMetric = new promClient.Gauge({
    name: delayedMetricName,
    help: 'Number of delayed jobs',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
  });

  const activeMetric = new promClient.Gauge({
    name: activeMetricName,
    help: 'Number of active jobs',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
  });

  const waitingMetric = new promClient.Gauge({
    name: waitingMetricName,
    help: 'Number of waiting jobs',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL],
  });

  const durationMetric = new promClient.Summary({
    name: durationMetricName,
    help: 'Time to complete jobs',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL, STATUS_LABEL],
    maxAgeSeconds: 300,
    ageBuckets: 13,
  });

  const waitingDurationMetric = new promClient.Summary({
    name: waitingDurationMetricName,
    help: 'Time spent waiting for a job to run',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL, STATUS_LABEL],
    maxAgeSeconds: 300,
    ageBuckets: 13
  });

  const attemptsMadeMetric = new promClient.Summary({
    name: attemptsMadeMetricName,
    help: 'Job attempts made',
    labelNames: [QUEUE_NAME_LABEL, QUEUE_PREFIX_LABEL, STATUS_LABEL],
    maxAgeSeconds: 300,
    ageBuckets: 13
  });

  function recordJobMetrics(labels: {[key: string]: string}, status: JobStatus, job: bull.Job) {
    if (!job.finishedOn) {
      return;
    }

    const jobLabels = {
      [STATUS_LABEL]: status,
      ...labels
    }

    const jobDuration = job.finishedOn - job.processedOn!;
    durationMetric.observe(jobLabels, jobDuration);

    const waitingDuration = job.processedOn - job.timestamp;
    waitingDurationMetric.observe(jobLabels, waitingDuration);

    attemptsMadeMetric.observe(jobLabels, job.attemptsMade);
  }

  function start(queue: bull.Queue) {

    // @ts-ignore
    const keyPrefix = queue.keyPrefix.replace(/.*\{|\}/gi, '')

    const labels = {
      [QUEUE_NAME_LABEL]: queue.name,
      [QUEUE_PREFIX_LABEL]: keyPrefix,
    }

    if (useGlobal) {
      queue.on('global:completed', async (jobId: number) => {
        const job = await queue.getJob(jobId);
        recordJobMetrics(labels, JobStatus.COMPLETED, job);
      });
      queue.on('global:failed', async (jobId: number) => {
        const job = await queue.getJob(jobId);
        recordJobMetrics(labels, JobStatus.FAILED, job)
      })
    } else {
      queue.on('completed', (job) => {
        recordJobMetrics(labels, JobStatus.COMPLETED, job);
      });
      queue.on('failed', (job) => {
        recordJobMetrics(labels, JobStatus.FAILED, job)
      })
    }

    const metricInterval = setInterval(() => {
      queue
        .getJobCounts()
        .then(({ completed, failed, delayed, active, waiting }) => {
          completedMetric.set(labels, (completed || 0));
          failedMetric.set(labels, (failed || 0));
          delayedMetric.set(labels, (delayed || 0));
          activeMetric.set(labels, (active || 0));
          waitingMetric.set(labels, (waiting || 0));
        });
    }, interval);

    const removeMetrics = () => {
      durationMetric.remove(labels);
      completedMetric.remove(labels);
      failedMetric.remove(labels);
      delayedMetric.remove(labels);
      activeMetric.remove(labels);
      waitingMetric.remove(labels);
      waitingDurationMetric.remove(labels);
      attemptsMadeMetric.remove(labels);
    }
    return {
      stop: () => clearInterval(metricInterval),
      remove: removeMetrics
    };
  }

  return {
    start,
  };
}
